### PR TITLE
ASoC: SOF: log compiler name and version information

### PR DESCRIPTION
**Pull request by ktrzcinx**
This PR is related with https://github.com/thesofproject/sof/pull/2108 and issue https://github.com/thesofproject/sof/issues/1678

### CODE REVIEW

File: include/sound/sof/info.h [hunk 1]
```diff
@@ -43,6 +54,7 @@ struct sof_ipc_fw_version {
 	uint8_t time[10];
 	uint8_t tag[6];
 	uint32_t abi_version;
+	struct sof_ipc_cc_version compiler;
```
Comment (plbossart) [inline include/sound/sof/info.h]: doesn't this break backwards compatibility?

**Comment by dbaluta** [comment]
@ktrzcinx this looks to be common code.Please remove 'Intel' from commit message.

File: sound/soc/sof/loader.c [hunk 2]
```diff
@@ -520,7 +549,7 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 		if (ret < 0) {
 			dev_err(sdev->dev, "error: snd_sof_debugfs_buf_item failed\n");
 			return ret;
-		}
+		}
```
Comment (paulstelian97) [inline sound/soc/sof/loader.c]: Extra whitespace?

Comment (ktrzcinx) [reply sound/soc/sof/loader.c]: deleted

File: sound/soc/sof/loader.c [hunk 3]
```diff
@@ -32,6 +32,31 @@ static int get_ext_windows(struct snd_sof_dev *sdev,
...
 
+static int get_cc_info(struct snd_sof_dev *sdev,
+		       struct sof_ipc_ext_data_hdr *ext_hdr)
+{
+	int ret;
+
+	struct sof_ipc_cc_version *cc =
+		container_of(ext_hdr, struct sof_ipc_cc_version, ext_hdr);
+
+	dev_info(sdev->dev, "Firmware info: used compiler %s %d:%d:%d%s used optimization flags %s\n",
```
Comment (plbossart) [inline sound/soc/sof/loader.c]: dev_dbg() ?

dev_info will be reported on most platforms, and it's noise for the average user

File: sound/soc/sof/loader.c [hunk 4]
```diff
@@ -65,6 +90,9 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 		case SOF_IPC_EXT_WINDOW:
 			ret = get_ext_windows(sdev, ext_hdr);
 			break;
+		case SOF_IPC_EXT_CC_INFO:
+			ret = get_cc_info(sdev, ext_hdr);
+			break;
```
Comment (plbossart) [inline sound/soc/sof/loader.c:106]: shouldn't we have a check that this is only provided once? Otherwise you'll have a warning with the debugfs item created multiple times.

File: sound/soc/sof/loader.c [hunk 5]
```diff
@@ -75,8 +75,7 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 
 	while (ext_hdr->hdr.cmd == SOF_IPC_FW_READY) {
 		/* read in ext structure */
-		offset += sizeof(*ext_hdr);
-		snd_sof_dsp_block_read(sdev, bar, offset,
+		snd_sof_dsp_block_read(sdev, bar, offset + sizeof(*ext_hdr),
```
Comment (plbossart) [inline sound/soc/sof/loader.c]: this should probably be a separate PR, so that we can properly test it. It really has nothing to do with the compiler.

Comment (ktrzcinx) [reply sound/soc/sof/loader.c]: Without this fix, my changes won't work.

File: sound/soc/sof/loader.c (hunk 3 above)
Comment (ktrzcinx) [reply sound/soc/sof/loader.c]: Ok

File: sound/soc/sof/loader.c (hunk 4 above)
Comment (ktrzcinx) [reply sound/soc/sof/loader.c:106]: ok

**Comment by ktrzcinx** [comment]
Whole extended FW boot data structure parsing is enclosed in such a loop:
```
while (ext_hdr->hdr.cmd == SOF_IPC_FW_READY) {
  [...]
  switch (ext_hdr->type) {
    [run action]
  }
  [read next header]
}
```
So when firmware (because of version or configuration) don't send structure with compiler information then get_cc_info action just won't run and that's it.
When we get older kernel (without support this extended header) and new FW then should be dev_dbg log with found header type and size, one more dev_warn with information about unknown ext header and no more troubles (after introducing my fix).

File: sound/soc/sof/loader.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/loader.c]: ok, will upstream this first then

**Comment by plbossart** [comment]
@xiulipan @fredoh9 does the checkpatch test work? I see warnings but the test result is ok?

File: sound/soc/sof/loader.c [hunk 6]
```diff
@@ -32,6 +32,33 @@ static int get_ext_windows(struct snd_sof_dev *sdev,
...
+	/* create read-only cc_version debugfs to store compiler version info */
+	if (sdev->first_boot) {
+		ret = snd_sof_debugfs_buf_item(sdev, cc, cc->ext_hdr.hdr.size,
+					"cc_version", 0444);
+
+		/* errors are only due to memory allocation, not debugfs */
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: snd_sof_debugfs_buf_item failed\n");
+			return ret;
+		}
```
Comment (ranj063) [inline sound/soc/sof/loader.c:65]: @ktrzcinx dont you need to save the cc info somewhere? the debugfs entry points to the address which will be freed after the extended data is parsed no?

Comment (ktrzcinx) [reply sound/soc/sof/loader.c:65]: Well, there is no such an information in debugfs description (and functions below) and test on real machine show that file has been written correctly

Comment (ranj063) [reply sound/soc/sof/loader.c:65]: @ktrzcinx take a look at sof_dfsentry_read(). This function tried to mopy the data from the df->buf which points to the base address that is passed when the dfsentry is created. In this case it is pointing to the ext_data that is passed to get_cc_info() which is eventually freed when the snd_sof_fw_parse_ext_data() function quits.

Comment (ktrzcinx) [reply sound/soc/sof/loader.c:65]: Ok, done

Comment (ktrzcinx) [reply sound/soc/sof/loader.c:65]: Ok, done

**Comment by plbossart** [comment]
@ktrzcinx build failed?

File: sound/soc/sof/loader.c [hunk 7]
```diff
@@ -32,6 +32,44 @@ static int get_ext_windows(struct snd_sof_dev *sdev,
...
+		 cc->optim);
+	
+
+	/* create read-only cc_version debugfs to store compiler version info */
+	/* use local copy of the cc_version to prevent data corruption */
+	if (sdev->first_boot) {
+		sdev->cc_version = devm_kmalloc(cdev->dev, cc->ext_hdr.hdr.size,
+						GFP_KERNEL);
+		
+		if(!sdev->cc_version) {
```
Comment (ranj063) [inline sound/soc/sof/loader.c]: sdev->cc_version was never defined?

Comment (ktrzcinx) [reply sound/soc/sof/loader.c]: Yea, I forgot one file during 'git add', my fault

File: include/sound/sof/info.h [hunk 8]
```diff
@@ -115,4 +116,18 @@ struct sof_ipc_window {
...
+	uint32_t major;
+	uint32_t minor;
+	uint32_t micro;
+
+	/* reserved for future use */
+	uint32_t reserved[4];
+
+	char name[16]; /* null ended compiler name */
+	char optim[4]; /* null ended compiler -O flag value */
+	char desc[]; /* null ended compiler description */
```
Comment (jajanusz) [inline include/sound/sof/info.h]: comments differ from these on fw side

Comment (ktrzcinx) [reply include/sound/sof/info.h]: updated

File: include/sound/sof/info.h (hunk 1 above)
Comment (ktrzcinx) [reply include/sound/sof/info.h]: outdated

**Comment by jajanusz** [comment]
> LGTM but I'd like a firmware person to verify. @jajanusz you good with this?

Yes, FW side part is already merged.
